crates/sui-graphql-rpc/src/extensions/timeout.rs

```diff
@@ -2,35 +2,69 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use async_graphql::{
-    extensions::{Extension, ExtensionContext, ExtensionFactory, NextRequest},
-    Response, ServerError,
+    extensions::{Extension, ExtensionContext, ExtensionFactory, NextExecute, NextParseQuery},
+    parser::types::ExecutableDocument,
+    Response, ServerError, ServerResult,
 };
-use std::sync::Arc;
+use async_graphql_value::Variables;
 use std::time::Duration;
-use tokio::time::timeout;
+use std::{net::SocketAddr, sync::Arc};
+use tokio::{sync::Mutex, time::timeout};
+use tracing::error;
+use uuid::Uuid;
 
-use crate::config::ServiceConfig;
+use crate::{config::ServiceConfig, error::code};
 
-#[derive(Clone, Debug, Default)]
-pub(crate) struct Timeout;
+#[derive(Debug, Default)]
+pub(crate) struct Timeout {
+    pub query: Mutex<Option<String>>,
+}
 
 impl ExtensionFactory for Timeout {
     fn create(&self) -> Arc<dyn Extension> {
-        Arc::new(Timeout)
+        Arc::new(Timeout {
+            query: Mutex::new(None),
+        })
     }
 }
 
 #[async_trait::async_trait]
 impl Extension for Timeout {
-    async fn request(&self, ctx: &ExtensionContext<'_>, next: NextRequest<'_>) -> Response {
+    async fn parse_query(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        query: &str,
+        variables: &Variables,
+        next: NextParseQuery<'_>,
+    ) -> ServerResult<ExecutableDocument> {
+        let document = next.run(ctx, query, variables).await?;
+        *self.query.lock().await = Some(ctx.stringify_execute_doc(&document, variables));
+        Ok(document)
+    }
+
+    async fn execute(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        operation_name: Option<&str>,
+        next: NextExecute<'_>,
+    ) -> Response {
         let cfg = ctx
             .data::<ServiceConfig>()
             .expect("No service config provided in schema data");
+        let query = self.query.lock().await.take().unwrap_or("".to_string());
         let request_timeout = Duration::from_millis(cfg.limits.request_timeout_ms);
-
-        timeout(request_timeout, next.run(ctx))
+        timeout(request_timeout, next.run(ctx, operation_name))
```

Comment:
IIUC `execute` is called once per top level query and a query can include many fields/subqueries. If a query times out, we won't know where exactly (i.e. while resolving which field) the timeout happens. I'm wondering if it makes sense to add the `timeout` and logging to `resolve` instead. This way we have the path information as well and a more precise target for optimization.

Reply:
Happy to have it either way. My thought was that it might make more sense to know all the query, such that we can tear it down to parts that do not work during our debugging. It might give us a bit more context of what is the need and what is the user trying to achieve, rather than just fixing having a sub-query that times out.

Reply:
Sounds good!